PR #16762 causes a reduction of precision to 16 significant digits in certain cases

zachetienne opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #17756 

#### Brief description of what is fixed or changed
The line of code `big = 1.0*abs(i)` in commit
https://github.com/sympy/sympy/commit/c1e25a445f1c27829ea8cf0e04dbdd6d749834ce

in certain cases reduced the number of digits of significance to 16. 

A workaround was found for the `big = 1.0*abs(i)` line of code, which does not reduce digits of significance.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

isuruf commented on sympy/core/tests/test_power.py:
```diff
@@ -467,14 +468,26 @@ def test_issue_2993():
     assert str((-2.3*x + 2)**0.3) == '1.28386201800527*(0.869565217391304 - x)**0.3'
     assert str((2.3*x + 2)**0.3) == '1.28386201800527*(x + 0.869565217391304)**0.3'
-    assert str((2.3*x - 4)**Rational(1, 3)) == '1.5874010519682*(0.575*x - 1)**(1/3)'
+    assert str((2.3*x - 4)**Rational(1, 3)) == '2**(2/3)*(0.575*x - 1)**(1/3)'
```
I'm not sure if this is desired. Multiplying by 1.0 was meant to fix this, but it had the unintended consequence of losign digits. To preserve significant digits, I think we should multiply big by `Float(1, dps=max_dps)` where `max_dps` is the maximum number of digits in all of the floats found in the add.

zachetienne replied:
Neat suggestion! However I'm having trouble implementing what you suggest. Here's as far as I was able to make it:

```python
        elif e.is_Number and abs(e) != 1:
            # handle the Float case: (2.0 + 4*x)**e -> 2.**e*(1 + 2.0*x)**e
            c, m = zip(*[i.as_coeff_Mul() for i in self.args])
            # vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
            max_dps = -1
            if any(i.is_Float for i in c):  # XXX should this always be done?
                max_dps = max(prec_to_dps(i._prec), max_dps)
            # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
            if any(i.is_Float for i in c):  # XXX should this always be done?
                big = -1
                for i in c:
                    if abs(i) >= big:
            # vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
                        big = Float(1,dps=max_dps)*abs(i)
            # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                if big > 0 and big != 1:
                    from sympy.functions.elementary.complexes import sign
                    bigs = (big, -big)
                    c = [sign(i) if i in bigs else i/big for i in c]
                    addpow = Add(*[c*m for c, m in zip(c, m)])**e
                    return big**e*addpow
```

I get the error

```
  File "............/add.py", line 393, in _eval_power
    max_dps = max(prec_to_dps(i._prec), max_dps)
AttributeError: 'Mul' object has no attribute '_prec'
```

smichr replied:
I would suggest to just keep it simple. If the largest number is a Rational then a Rational will get extracted. No harm done and the expression is more compact.

smichr replied:
otherwise, would this do it:
```python
elif e.is_Number and abs(e) != 1:
    # handle the Float case: (2.0 + 4*x)**e -> 2.**e*(1 + 2.0*x)**e
    c, m = zip(*[i.as_coeff_Mul() for i in self.args])
    if any(i.is_Float for i in c):  # XXX should this always be done?
        big = -1
        for i in c:
            if abs(i) >= big:
                big = abs(i)
        if big > 0 and big != 1:
            from sympy.functions.elementary.complexes import sign
            bigs = (big, -big)
            c = [sign(i) if i in bigs else i/big for i in c]
            addpow = Add(*[c*m for c, m in zip(c, m)])**e
            max_dps = -1
            for i in c:
                if i.is_Float:
                    max_dps = max(prec_to_dps(i._prec), max_dps)
            if max_dps != -1:
                big = Float(1, dps=max_dps)*big
            return big**e*addpow
```

zachetienne replied:
@smichr : Hmm... I get the following error when trying to run with your latest patch:

```
  File "/............/sympy/core/add.py", line 402, in _eval_power
    for i in c:
TypeError: 'Float' object is not iterable
```

isuruf replied:
```
addpow = Add(*[d*m for d, m in zip(c, m)])**e
```

zachetienne replied:
@isuruf : Thanks that fixed the "not iterable" error. I had to update the two unit tests that were obviously affected. Please take a look and let us know if this seems acceptable. 

Thanks so much!

isuruf commented on sympy/core/tests/test_power.py:
```diff
@@ -469,12 +470,24 @@ def test_issue_2993():
     assert str((2.3*x - 4)**Rational(1, 3)) == '1.5874010519682*(0.575*x - 1)**(1/3)'
     eq = (2.3*x + 4)
-    assert eq**2 == 16.0*(0.575*x + 1)**2
+    assert str(eq**2) == '16.0*(0.575*x + 1)**2'
```
```suggestion
    assert eq**2 == 16.0*(0.575*x + 1)**2
```

smichr commented:
A more-minimum failing example is
```python
len(sqrt(x/(x + y)**2 + Float('0.008', 30)).subs(y, pi.n(25)).atoms(Float))
```
This gives 2 in master but 3 in this PR. I think the answer should be 2.

zachetienne commented:
@smichr: 

> This gives 2 in master but 3 in this PR. I think the answer should be 2.

After the subs(), I get
```
     ___________________________________________
    ╱                 x                         
   ╱  ───────────────────────────────── + 0.008 
  ╱                                   2         
╲╱    (x + 3.141592653589793238462643)          

```
in master, and
```
                                      __________________________________________________________________________
                                     ╱                  x                                                       
0.318309886183790671537767525692⋅   ╱  ──────────────────────────────────── + 0.0789568352087148689506759285212 
                                   ╱                                      2                                     
                                 ╲╱    (0.3183098861837906715377675⋅x + 1)                                      
```
after this PR. Both seem plausibly identical. Are you suggesting that you aren't fully satisfied with your patch? If not, is there an earlier version you would prefer, or something else?

smichr commented:
This diff gives 2 constants and still passes the added tests:
```python
diff --git a/sympy/core/add.py b/sympy/core/add.py
index 933500c..3aab4de 100644
--- a/sympy/core/add.py
+++ b/sympy/core/add.py
@@ -2,7 +2,6 @@
 
 from collections import defaultdict
 from functools import cmp_to_key
-from mpmath.libmp import prec_to_dps
 
 from .basic import Basic
 from .compatibility import reduce, is_sequence, range
@@ -11,7 +10,7 @@
 from .singleton import S
 from .operations import AssocOp
 from .cache import cacheit
-from .numbers import ilcm, igcd, Float
+from .numbers import ilcm, igcd
 from .expr import Expr
 
 # Key for sorting commutative args in canonical order
@@ -398,12 +397,6 @@ def _eval_power(self, e):
                     bigs = (big, -big)
                     c = [sign(i) if i in bigs else i/big for i in c]
                     addpow = Add(*[d*m for d, m in zip(c, m)])**e
-                    max_dps = -1
-                    for i in c:
-                        if i.is_Float:
-                            max_dps = max(prec_to_dps(i._prec), max_dps)
-                    if max_dps != -1:
-                        big = Float(1, dps=max_dps)*big
                     return big**e*addpow
     @cacheit
     def _eval_derivative(self, s):
```
The test that there only be 2 atoms should be added.

zachetienne commented:
@smichr : Indeed it makes the sqrt expression simpler. However, not all unit tests pass; assertions

```
assert str((2.3*x - 4)**Rational(1, 3)) == '1.5874010519682*(0.575*x - 1)**(1/3)'
assert eq**2 == 16.0*(0.575*x + 1)**2
```
fails, with the left-hand-sides of the equalities evaluating again to
```
2**(2/3)*(0.575*x - 1)**(1/3)
16*(0.575*x + 1)**2
```
If I were given the choice between this behavior or the 3-atom Float behavior of the sqrt() expression above, I'd choose this one. Of course it's an aesthetic judgement.

smichr commented:
> Of course it's an aesthetic judgement.

It is also a matter of accuracy. If you extract a Rational as a Float it will not behave the same when the expression is fully substituted with numerical values.

zachetienne commented:
@smichr : Fair point. Normally (in the world of 64-bit double precision arithmetic) I'd expect `2.3` to only be known to some fixed precision, so factoring out 2 wouldn't help. It is possible this might help the precision. Again, I approve the current patch, for what it's worth. It fixes the original issue I was seeing and yields good results.

smichr commented on sympy/core/add.py:
```diff
@@ -387,19 +387,17 @@ def _eval_power(self, e):
         elif e.is_Number and abs(e) != 1:
             # handle the Float case: (2.0 + 4*x)**e -> 2.**e*(1 + 2.0*x)**e
```
This should be `(2.0 + 4*x)**e -> 4**e*(0.5 + x)**e`

zachetienne replied:
Done.

smichr replied:
I made one more comment about the issue test case. Other than that it looks like the issue is resolved. Thank for your work on this.

smichr commented on sympy/core/tests/test_power.py:
```diff
@@ -467,14 +468,28 @@ def test_issue_2993():
+    r_at_max_density,x1,x8,x10,x11,x12 = symbols('r_at_max_density x1 x8 x10 x11 x12',real=True)
+    expr = sqrt(1 + N('0.0170992456333788667034850458615',30)*(r_at_max_density**4 + x1*x10 - 2*x1*x11 - x12*(-x1 + x10))**2/(r_at_max_density**3*x8**2*(x10 - 3*x11 + 2*x12)**2))/2
+    expr = expr.subs(x8, N('0.0966594956075474769169134801223',30))
+    ## ^ above subs() is where issue #17756 loss of significance occurs.
```
I actually don't see the difference between this PR and master at this step. Suggest:
```python
# issue 17756
a, b, c, d, e, f = symbols('a:f')
expr = sqrt(1 + a*(c**4 + g*d - 2*g*e - f*(-g + d))**2/
    (c**3*b**2*(d - 3*e + 2*f)**2))/2
r = [
(a, N('0.0170992456333788667034850458615', 30)),
(b, N('0.0966594956075474769169134801223', 30)),
(c, N('0.390911862903463913632151616184', 30)),
(d, N('0.152812084558656566271750185933', 30)),
(e, N('0.137562344465103337106561623432', 30)),
(f, N('0.174259178881496659302933610355', 30)),
(g, N('0.220745448491223779615401870086', 30))]
tru = expr.n(30, subs=dict(r))
seq = expr.subs(r)
# although `tru` is the right way to evaluate
# expr with numerical values, `seq` will have
# significant loss of precision if extraction of
# the largest coefficient of a power's base's terms
# is done improperly
assert seq == tru
```

zachetienne replied:
I like it. Much cleaner, and does the job well (assertion fails in master but passes beautifully after this PR). Committed.

smichr replied:
If you decide to clean up the history with local commands
```
git diff 78d5a35 > dif
git co 78d5a35
git branch -b tmp
git apply dif
```
Then you could modify the line were addpow is calculated to
```
addpow = Add(*[c*m for c, m in zip(c, m)])**e
```
(no need to rename c to d). And then commit with a single commit with a short title and push it here with the `force` option.
```
git commit -am 'your short title'
git branch -D master
git co -b master
git branch -D tmp
git push -f zachetienne master
```

zachetienne replied:
@smichr : Okay, I followed your instructions as best I could (co == checkout, etc). I think I did the correct thing. Please take a look and let me know. Thanks!

smichr replied:
Oops (re `co`). And yes, it looks good. But if you reedit the commit message and put a blank line after the title then you won't get the whole commit message as the title.

```
Fix issue #17756

PR #16762 causes a reduction of precision to 16 significant digits. 

Special thanks to Christopher Smith (@smichr) and Isuru Fernando (@isuruf)
for their valuable input and contributions to this patch.
```

zachetienne replied:
Okay, I think that (git commit --amend) did it.

zachetienne committed:
Fix issue #17756

PR #16762 causes a reduction of precision to 16 significant digits.

Special thanks to Christopher Smith (@smichr) and Isuru Fernando (isuruf) for their valuable input and contributions to this patch.

zachetienne committed:
Fix issue #17756: PR #16762 causes a reduction of precision to 16 significant digits. Special thanks to Christopher Smith (@smichr) and Isuru Fernando (isuruf) for their valuable input and contributions to this patch.

zachetienne committed:
Merge branch 'master' of github.com:zachetienne/sympy

smichr commented:
OK, seeing no feedback I will commit this.
